lib/src/json/builder/builder.dart

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import "../reader/reader.dart";
+
+/// A function reading JSON data from a [JsonReader] into a Dart value.
+///
+/// This is a general type which can be implemented
+/// for any type that can be built from JSON data.
+typedef JsonBuilder<T> = T Function(JsonReader);
+
+/// Parses a JSON value from [reader] into a plain JSON-like structure.
+///
+/// A JSON-like structure is either a number, a string, a boolean, `null`, or
+/// a `List<dynamic>` containing JSON-like structures, or a `Map<String, dynamic>`
+/// where the values are JSON-like structures.
+dynamic jsonValue(JsonReader reader) {
```

Comment:
Can we consider `Object` (or `Object?` soon) for these APIs so that it isn't as easy to misuse the return value?

Reply:
Probably a good idea.
Dynamic is dangerous.